### PR TITLE
fix batch normalization

### PR DESCRIPTION
**Pull request by the-moliver**
The previous implementation of batch norm didn't match the paper or other implementations, i.e. https://github.com/torch/nn/blob/master/BatchNormalization.lua
It didn't even have different behavior for training and test time!
I set the new version to be the default, but left the previous ones for backward compatibility. The running estimates of the mean and std are now only used at test time. I also set the default momentum to 0.1 as in the torch implementation. Hopefully this helps some of the issues people have had with batch norm...

### CODE REVIEW

**Comment by fchollet** [comment]
I just checked with Christian about the correct way to do batch-norm. You're right. But that's not all, we should also normalize per feature map rather than per unit.

So here's what we should do:
- keep just 2 modes, 1 and 2
- change mode 1 to use the batch statistics rather than the running averages
- normalize per feature map (last dimension) rather than per unit.

Can you update your PR with these changes?

**Comment by fchollet** [comment]
Wait, actually, because we support both Theano-style 4D tensors and TensorFlow-style 4D tensors, we should make the axis of feature maps configurable. For instance with `dim_ordering="th"`, the feature maps are on the 2nd axis, but for `dim_ordering="tf"`, they're on the 4th axis (which is saner, btw).

So let's have an "axis" argument defaulting to -1. And let's explain this gotcha in the docs.

**Comment by fchollet** [comment]
Never mind, I took care of it.

**Comment by AvantiShri** [comment]
Thank you for this fix. One lingering question I can't quite figure out from the code: what setting of mode/axis should be used for batch normalising fully-connected layers in accordance with the batch normalisation paper?

**Comment by fchollet** [comment]
@AvantiShri for 2D tensors, the default of -1 will work fine (or equivalently, 1).

**Comment by AvantiShri** [comment]
Ah, because each unit in the layer is treated as a separate "channel". makes sense, thank you.

**Comment by sabersword** [comment]
Follow up in the case of CNN, TensorFlow documentation for [tf.nn.moments](https://www.tensorflow.org/api_docs/python/tf/nn/moments) states:
When using these moments for batch normalization
* for so-called **global normalization**, used with convolutional filters with shape [batch, height, width, depth], pass axes=[0, 1, 2].
* for simple batch normalization pass axes=[0] (**batch only**).

When keras uses tf backend, BatchNormalization resorts to "global" mode by default. How can we use the "batch only" mode mentioned above? The way the [code](https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py) is structured, I cannot see how the proper reduction axes can be passed to Keras BatchNormalization.
